lib/java_buildpack/util/sanitizer.rb

```diff
@@ -22,11 +22,32 @@ class String
   #
   # @return [String] the sanitized uri
   def sanitize_uri
+    keywords = /key|password|username|cred[entials]*[s]*|password|token|api[-_]token|api|auth[entication]*|access[-_]token|secret[-_]token/i
+
     rich_uri          = URI(self)
     rich_uri.user     = nil
     rich_uri.password = nil
-    rich_uri.query = rich_uri.query&.gsub(/(Api-Token=dt\w*\.\w*)\.\w*/, '\1.REDACTED')
+
+    if(rich_uri.query)
+      params = Hash[URI.decode_www_form rich_uri.query]
+
+      query_params = ""
+
+      params.each do |key,value|
+        match = key.match(keywords)
+
+        if(match)
+          if(match[0] == "Api-Token" && value =~ /dt\w*/)
+            params[key] = value.gsub(/(dt\w*\.\w*)\.\w*/, '\1.REDACTED')
```

Comment:
What is the need for this inner `if` block? Why can't we just have everything be masked in the same way?

Reply:
Thanks for your review. The if is needed because we only want to mask parts of the Dynatrace API Token. The Tokens have a public and a private part. Masking only the private part helps in identifying which token was used for staging the app by looking at the logs.

Reply:
Just one note from my side. My concern is that if we manipulate the URI we might end up with similar Hashes for different files, possibly breaking other stuff in the Buildpack. We are touching the Buildpack Code at a very low level here. This Buildpack is a very complex piece of software and we should be aware of possible side effects.

The sanitize_uri function is used in different places, mainly for Logging URL's and generating a filename for the cached file artifacts. (File, Etag, Last-Modified)

The Download Cache itself is used for a lot of things. Main Use Case is for downloading Stuff directly by using the download function in the Base Component but also in some other components like External Config or the Repository Index. 

But only someone with a lot of knowledge about all the mechanisms in the Buildpack can judge possible side effects. I can't.

The pull request is also missing the tests.

Reply:
Hi @dmikusa,
I've updated the code to fulfill the RuboCop recommendations and also added a test that includes the tests. Please have another look. Thanks in advance.

Reply:
Thanks for this PR & for bearing with me through the process.

I'm not a huge fan of having the special case part in there, but mostly because I think it could be potentially difficult to handle. If there's one special case, then that invites the potential for others. We can cross that bridge if it becomes an issue though.

>The sanitize_uri function is used in different places, mainly for Logging URL's and generating a filename for the cached file artifacts. (File, Etag, Last-Modified)

This is a good point. If it's caching a URL with the masked value, that would be less unique and could introduce collisions (i.e. the masked value changes, but it's not reflected in the URL for caching purposes). 

Since the buildpack cache is scoped to your application, I'm not sure this would be a huge issue. Taking Dynatrace for example. The user would need to perform a build with one key and then switch the key and rebuild. In that case, I think you'd get a collision. It's hard to say if that would be a problem. In most cases where everyone gets the same set of binaries, it wouldn't be. If the binary being fetched was customized and unique to the particular URL then that could be an issue.

I could see this being handled in two ways.

1. We could adjust the caching to pull the real URL, not the masked URL, that way we have the full and unique URL for caching purposes. The trick with this approach is that we'd need to be really careful to not have the actual URL saved somewhere in the cache as that would be another form of information leak.
2. Since most dependencies are not impacted by this, as they are not unique per app/user/token, we could handle this for the particular framework. Essentially, the framework having domain knowledge would know if the dependency might be susceptible to this problem. It could then clear the cache when it detects a change to the secret/token/password/etc.

At any rate, I'm not sure if we need to deal with this now or if we can wait and follow up with another PR. I'd defer to you on that, given you know how this might impact Dynatrace users. What do you think?

Reply:
@dmikusa 
I agree that the special handling of the DT Token in that place is not a good idea. I suggest removing that part again. Then we have at least the problem with logging secrets solved for now.